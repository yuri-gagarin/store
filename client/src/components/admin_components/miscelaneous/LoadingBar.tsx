import React from "react";
import { Dimmer, Loader, Segment } from "semantic-ui-react";

const LoadingBar: React.FC<{}> = (props): JSX.Element => {

  return (
    <div style={{ height: "50px", width: "100%" }}>
      <Segment>
        <Dimmer active>
          <Loader>Loading</Loader>
        </Dimmer>
      </Segment>
    </div>
  );
};

export default LoadingBar;
