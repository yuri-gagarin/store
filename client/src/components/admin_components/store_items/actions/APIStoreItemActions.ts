import { Dispatch } from "react";
import axios, { AxiosRequestConfig, AxiosError } from "axios";
// state //
import { IGlobalAppState } from "../../../../state/Store";
// type definitions //
import { 
  StoreItemQueryPar, IStoreItemServerResData, IStoreItemServerRes,
  IStoreItemImgServerResData, IStoreItemImgServerRes,
  ClientStoreItemData
} from "../type_definitions/storeItemTypes";


export const getAllStoreItems = (dispatch: Dispatch<StoreItemAction>, queryParams?: StoreItemQueryPar): Promise<void> => {
  const requestOptions: AxiosRequestConfig = {
    method: "get",
    url: "/api/store_items",
    params: queryParams
  };
  dispatch({ type: "DISPATCH_STORE_ITEM_API_REQUEST", payload: { loading: true } });
  return axios.request<IStoreItemServerResData, IStoreItemServerRes>(requestOptions)
    .then((response) => {
      const { data } = response;
      const storeItems = data.storeItems!;
      dispatch({ type: "GET_ALL_STORE_ITEMS", payload: {
        loading: false,
        responseMsg: data.responseMsg,
        numberOfItems: data.numberOfItems ? parseInt(data.numberOfItems, 10) : undefined,
        loadedStoreItems: storeItems,
        error: null
      }});
      return Promise.resolve();
    })
    .catch((error: AxiosError) => {
      dispatch({ type: "SET_STORE_ITEM_ERROR", payload: {
        loading: false,
        responseMsg: error.message,
        error: error
      }});
      return Promise.reject();
    });
};

export const getStoreItem = (_id: string, dispatch: Dispatch<StoreItemAction>): Promise<void> => {
  const requestOptions: AxiosRequestConfig = {
    method: "get",
    url: "/api/store_items/" + _id,
  };
  dispatch({ type: "DISPATCH_STORE_ITEM_API_REQUEST", payload: { loading: true } });
  return axios.request<IStoreItemServerResData, IStoreItemServerRes>(requestOptions)
    .then((response) => {
      const { data } = response;
      const storeItem = data.storeItem!;
      dispatch({ type: "GET_STORE_ITEM", payload: {
        loading: false,
        responseMsg: data.responseMsg,
        currentStoreItemData: storeItem,
        error: null
      }});
      return Promise.resolve();
    })
    .catch((error: AxiosError) => {
      dispatch({ type: "SET_STORE_ITEM_ERROR", payload: {
        loading: false,
        responseMsg: error.message,
        error: error
      }});
      return Promise.reject();
    });
};

export const createStoreItem = (data: ClientStoreItemData, dispatch: Dispatch<StoreItemAction>): Promise<void> => {
  console.log(72)
  const requestOptions: AxiosRequestConfig = {
    method: "post",
    url: "/api/store_items/create",
    data: data
  };
  dispatch({ type: "DISPATCH_STORE_ITEM_API_REQUEST", payload: { loading: true } });
  console.log("dispatchhed")
  return axios.request<IStoreItemServerResData, IStoreItemServerRes>(requestOptions)
    .then((response) => {
      const { data } = response;
      const newStoreItem = data.newStoreItem!
      console.log(data)
      dispatch({ type: "CREATE_STORE_ITEM", payload: {
        loading: false,
        responseMsg: data.responseMsg,
        newStoreItem: newStoreItem,
        error: null
      }});
      return Promise.resolve();
    })
    .catch((error: AxiosError) => {
      dispatch({ type: "SET_STORE_ITEM_ERROR", payload: {
        loading: false,
        responseMsg: error.message,
        error: error
      }});
      return Promise.reject();
    });
};

export const editStoreItem = (_id: string, data: ClientStoreItemData, dispatch: Dispatch<StoreItemAction>, state: IGlobalAppState): Promise<void> => {
  const { loadedStoreItems } = state.storeItemState;
  const requestOptions: AxiosRequestConfig = {
    method: "patch",
    url: "/api/store_items/update/" + _id,
    data: data
  };
  dispatch({ type: "DISPATCH_STORE_ITEM_API_REQUEST", payload: { loading: true } });
  return axios.request<IStoreItemServerResData, IStoreItemServerRes>(requestOptions)
    .then((response) => {
      const { data } = response;
      const editedStoreItem = data.editedStoreItem!;
      const updatedStoreItems = loadedStoreItems.map((storeItem) => {
        if (storeItem._id === editedStoreItem._id) {
          return editedStoreItem;
        } else {
          return storeItem;
        }
      });
      dispatch({ type: "EDIT_STORE_ITEM", payload: {
        loading: false,
        responseMsg: data.responseMsg,
        editedStoreItem: editedStoreItem,
        loadedStoreItems: updatedStoreItems,
        error: null
      }});
      return Promise.resolve();
    })
    .catch((error: AxiosError) => {
      dispatch({ type: "SET_STORE_ITEM_ERROR", payload: {
        loading: false,
        responseMsg: error.message,
        error: error
      }});
      return Promise.reject();
    });
};

export const deleteStoreItem = (_id: string, dispatch: Dispatch<StoreItemAction>, state: IGlobalAppState): Promise<void> => {
  const { loadedStoreItems } = state.storeItemState;
  const requestOptions: AxiosRequestConfig = {
    method: "delete",
    url: "/api/store_items/delete/" + _id,
  };
  dispatch({ type: "DISPATCH_STORE_ITEM_API_REQUEST", payload: { loading: true } });
  return axios.request<IStoreItemServerResData, IStoreItemServerRes>(requestOptions)
    .then((response) => {
      const { data } = response;
      const deletedStoreItem = data.deletedStoreItem!;
      const updatedStoreItems = loadedStoreItems.filter((storeItem) => {
        return storeItem._id !== deletedStoreItem._id;
      })

      dispatch({ type: "DELETE_STORE_ITEM", payload: {
        loading: false,
        responseMsg: data.responseMsg,
        loadedStoreItems: updatedStoreItems,
        error: null
      }});
      return Promise.resolve();
    })
    .catch((error: AxiosError) => {
      dispatch({ type: "SET_STORE_ITEM_ERROR", payload: {
        loading: false,
        responseMsg: error.message,
        error: error
      }});
      return Promise.reject();
    });
};

export const uploadStoreItemImage = (storeItemId: string, imageFile: FormData, state: IGlobalAppState, dispatch: Dispatch<StoreItemAction>): Promise<void> => {
  const { loadedStoreItems } = state.storeItemState;
  const requestOptions: AxiosRequestConfig = {
    method: "post",
    url: "/api/uploads/store_item_images/" + storeItemId,
    data: imageFile
  };
  dispatch({ type: "DISPATCH_STORE_ITEM_API_REQUEST", payload: { loading: true } });
  return axios.request<IStoreItemImgServerResData, IStoreItemImgServerRes>(requestOptions)
    .then((response) => {
      const { data } = response;
      const { responseMsg, updatedStoreItem } = data;

      const updatedStoreItems = loadedStoreItems.map((storeItem) => {
        if (storeItem._id === updatedStoreItem._id) {
          return updatedStoreItem;
        } else {
          return storeItem;
        }
      });
      dispatch({ type: "UPLOAD_NEW_STORE_ITEM_IMG", payload: {
        loading: false,
        responseMsg: responseMsg,
        editedStoreItem: updatedStoreItem,
        loadedStoreItems: updatedStoreItems,
        error: null
      }});
      return Promise.resolve();
    })
    .catch((error: AxiosError) => {
      dispatch({ type: "SET_STORE_ITEM_ERROR", payload: {
        loading: false,
        responseMsg: error.message,
        error: error
      }});
      return Promise.reject();
    });
};

export const deleteStoreItemImage = (imgId: string, state: IGlobalAppState, dispatch: Dispatch<StoreItemAction>): Promise<void> => {
  const { loadedStoreItems } = state.storeItemState; 
  const { _id: storeItemId } = state.storeItemState.currentStoreItemData;

  const requestOptions: AxiosRequestConfig = {
    method: "delete",
    url: "/api/uploads/store_item_images/" + imgId + "/" + storeItemId
  };
  dispatch({ type: "DISPATCH_STORE_ITEM_API_REQUEST", payload: { loading: true } });
  return axios.request<IStoreItemImgServerResData, IStoreItemImgServerRes>(requestOptions)
    .then((response) => {
      const { data } = response;
      const { responseMsg, updatedStoreItem } = data;
      
      const updatedStoreItems = loadedStoreItems.map((storeItem) => {
        if (storeItem._id === updatedStoreItem._id) {
          return updatedStoreItem;
        } else {
          return storeItem;
        }
      });
      dispatch({ type: "DELETE_STORE_ITEM_IMG", payload: {
        loading: false,
        responseMsg: responseMsg,
        editedStoreItem: updatedStoreItem,
        loadedStoreItems: updatedStoreItems,
        error: null
      }});
      return Promise.resolve();
    })
    .catch((error: AxiosError) => {
      dispatch({ type: "SET_STORE_ITEM_ERROR", payload: {
        loading: false,
        responseMsg: error.message,
        error: error
      }});
      return Promise.reject();
    });
};


