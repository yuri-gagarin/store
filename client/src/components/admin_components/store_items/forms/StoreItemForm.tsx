import React, { useState, useEffect, useRef }  from "react";
import { Button, Form, TextArea } from "semantic-ui-react";
// additional components //
import StoreItemCategories from "./StoreItemCategories";
import StoreDetails from "./StoreDetails";
import StoreNameDropDown from "./StoreNameDropdown";
// actions and state //
import { IGlobalAppState, AppAction } from "../../../../state/Store";

export type FormState = {
  storeId: string;
  storeName: string;
  name: string;
  description: string;
  details: string;
  price: string;
  categories: string[];
}

interface Props {
  state: IGlobalAppState;
  dispatch: React.Dispatch<AppAction>;
  storeItem: IStoreItemData;
  handleCreateStoreItem(data: FormState): void;
  handleUpdateStoreItem(data: FormState): void;
}

const StoreItemForm: React.FC<Props> = ({ storeItem, handleCreateStoreItem, handleUpdateStoreItem, state, dispatch }): JSX.Element => {
  const { storeId, storeName, name, description, details, price, categories } = storeItem;
  const { currentStoreData } = state.storeState;
  // state and refs //
  const [ newForm, setNewForm ] = useState<boolean>(true)
  const [ formState, setFormState ] = useState<FormState>({ storeId, storeName, name, description, details, price, categories });
  const storeItemFormRef = useRef<HTMLDivElement>(document.createElement("div"));
  
  const handleTitleChange = (e: React.ChangeEvent<HTMLInputElement>):void => {
    setFormState({
      ...formState,
      name: e.target.value
    });
  };
  const hadnleDescriptionChange = (e: React.ChangeEvent<HTMLInputElement>): void => {
    setFormState({
      ...formState,
      description: e.target.value
    });
  };
  const handlePriceChange = (e: React.ChangeEvent<HTMLInputElement>): void => {
    setFormState({
      ...formState,
      price: e.target.value
    });
  };
  const handleDetailsChange = (e: React.ChangeEvent<HTMLInputElement>): void => {
    setFormState({
      ...formState,
      details: e.target.value
    });
  };
  const handleCategoryChange = (categories: string[]): void => {
    setFormState({
      ...formState,
      categories: [ ...categories ]
    });
  };

  const handleSubmit = (e: React.MouseEvent<HTMLButtonElement>) => {
    e.preventDefault();
    if (newForm) {
      handleCreateStoreItem(formState);
    } else {
      handleUpdateStoreItem(formState);
    }
  };

  useEffect(() => {
    // edit form //
    if (storeItem._id) {
      setNewForm(false);
    } else {
      setNewForm(true);
    }
  },  [storeItem]);

  useEffect(() => {
    if (storeItemFormRef.current) {
      const elem = storeItemFormRef.current.getBoundingClientRect();
      window.scrollTo({
        top: elem.bottom,
        behavior: "smooth"
      })
    }
  }, [storeItemFormRef]);
  
  return (
    <div className="createStoreItemFormHolder" ref={storeItemFormRef}>
      <Form id="createStoreItemForm">
        {
          (storeId && storeName) ? <StoreDetails storeId={storeId} storeName={storeName} createdAt={currentStoreData.createdAt}/> : null
        }
        <Form.Field>
          <label>Which Store to place the item in?</label>
          <StoreNameDropDown state={state} dispatch={dispatch} />
        </Form.Field>
        <Form.Field>
          <label>Store Item name</label>
          <input 
            id="storeItemFormTitle"
            onChange={handleTitleChange} 
            placeholder="Store Item name here ..." 
            value={formState.name}
          />
        </Form.Field>
        <Form.Field>
          <label>Store Item price</label>
          <input 
            id="storeItemFormPrice"
            onChange={handlePriceChange} 
            placeholder="Store Item price here..." 
            value={formState.price}
          />
        </Form.Field>
        <Form.Field>
          <label>Categories</label>
          <StoreItemCategories _handleCategoryChange={handleCategoryChange} />
        </Form.Field>
        <Form.Field
          id="storeItemFormDetails"
          control={TextArea}
          label='Store Item Details'
          onChange={handleDetailsChange}
          placeholder='Store Item details here...'
          value={formState.details}
         />
        <Form.Field
          id="storeItemFormDescription"
          control={TextArea}
          label='Store Item Description'
          onChange={hadnleDescriptionChange}
          placeholder='Store Item description here...'
          value={formState.description}
         />
         {
           newForm 
            ? <Button id="adminStoreItemFormCreate" type='submit' onClick={handleSubmit} content= "Create  New Store Item" />
            : <Button id="adminStoreItemFormUpdate" type='submit' onClick={handleSubmit} content= "Update Store Item" />
         }  
      </Form>
    </div>
  );
};

export default StoreItemForm;