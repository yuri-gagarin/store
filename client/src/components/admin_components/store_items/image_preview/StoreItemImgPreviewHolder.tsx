import React from "react";
// additional components //
import StoreItemImgThumb from "./StoreItemImgThumb";
// css  imports //
import "./css/storeItemImgPreviewHolder.css";
// state //
import { IGlobalAppState, AppAction } from "../../../../state/Store";
// helpers
import { deleteStoreItemImage } from "../actions/APIStoreItemActions";

// holder for image preview thumbnails //
interface Props {
  state: IGlobalAppState;
  dispatch: React.Dispatch<AppAction>
}
const StoreItemImgPreviewHolder: React.FC<Props> = ({ state, dispatch }): JSX.Element => {
  const { images } = state.storeItemState.currentStoreItemData;

  const handleDelete = (_id: string) => {
    deleteStoreItemImage(_id, state, dispatch);
  };

  if (images.length > 0) {
    return (
      <div id="storeItemImgPreviewHolder">
        <div className="storeItemImgHolderTitle">Uploaded Images</div>
        {
          images.map((image) => {
            return (
              <StoreItemImgThumb 
                key={image._id} 
                _id={image._id} 
                url={image.url} 
                handleDelete={handleDelete}
              />
            );
          })
        }
      </div>
    );
  } else {
    return (
      <div id="storeItemImgPreviewNone">
        <h3>No images uploaded... you should upload some... </h3>
      </div>
    );
  }
};

export default StoreItemImgPreviewHolder;


