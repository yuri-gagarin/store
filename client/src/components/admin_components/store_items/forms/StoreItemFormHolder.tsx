import React, { useContext, useState, useEffect } from "react";
import { Button, Grid, List } from "semantic-ui-react";
// css imports //
import "./css/storeItemFormHolder.css";
// additional components //
import StoreItemForm from "./StoreItemForm";
import StoreItemImgPreviewHolder from "../image_preview/StoreItemImgPreviewHolder";
import StoreItemImgUplForm from "./StoreItemImgUplForm";
import FormErrorComponent from "../../popups/FormErrorComponent";
import LoadingBar from "../../miscelaneous/LoadingBar";
// state //
import { Store } from "../../../../state/Store";
// client routing //
import { RouteComponentProps, withRouter } from "react-router-dom";
// api actions //
import { createStoreItem, editStoreItem } from "../actions/APIStoreItemActions";
import { getAllStores } from "../../stores/actions/APIstoreActions";
// ui actions //
import { openStoreItemForm, closeStoreItemForm } from "../actions/UIStoreItemActions";
// helpers //
import { ConvertDate } from "../../../helpers/displayHelpers";
import { checkSetValues } from "../../../helpers/validationHelpers";
// types 
import { AxiosError } from "axios";
import { StoreItemFormState } from "../type_definitions/storeItemTypes";

interface Props extends RouteComponentProps {

}

type StoreItemData = {
  storeId: string,
  storeName: string,
  name: string;
  description: string;
  details: string;
  price: string;
  images: IStoreItemImgData[];
  categories: string[];
}
type StoreDropdownData = {
  storeId: string;
  storeName: string;
}

const StoreItemFormHolder: React.FC<Props> = ({ history }): JSX.Element => {
  const { state, dispatch } = useContext(Store);
  const { loading, currentStoreItemData, storeItemFormOpen, error } = state.storeItemState;
  const { storeId, storeName, name, description, details, price, categories, createdAt, editedAt } = currentStoreItemData;
  // local component state //
  // const [ formOpen, setFormOpen ] = useState<boolean>(false);
  const [ newForm, setNewForm ] = useState<boolean>(true);
  const [ storeDropDownData, setStoreDropDownData ] = useState<StoreDropdownData[]>();
  
  // StoreItemForm toggle //
  const toggleForm = () => {
    storeItemFormOpen ? closeStoreItemForm(dispatch) : openStoreItemForm(dispatch);
  };
  // API call handlers CREATE - EDIT //
  const handleCreateStoreItem = ({ storeId, storeName, name, price, description, details, categories }: StoreItemFormState): void => {
    const storeItemData: StoreItemData = {
      storeId,
      storeName,
      name,
      description,
      details,
      price,
      images: currentStoreItemData.images,
      categories,
    };
    createStoreItem(storeItemData, dispatch)
      .then((_) => {
        // storeItem created //
        closeStoreItemForm(dispatch);
        // push to view Item
        // history.push("")
      })
      .catch((_) => {
        // handle error? show some modal? //
      });
  };
  const handleUpdateStoreItem = ({ storeId, storeName,name, description, details, price, categories }: StoreItemFormState): void => {
    const storeItemParams: StoreItemData = {
      storeId, storeName, name, details, description, price, images: currentStoreItemData.images, categories
    };
    editStoreItem(currentStoreItemData._id, storeItemParams, dispatch, state)
      .then((_) => {
        closeStoreItemForm(dispatch);
      })
      .catch((_) => {
        // handle error show modal ? //
      });
  };
  // lifecycle hooks //
  useEffect(() => {
    let mounted = true;
    if (mounted) {
      getAllStores(dispatch)
        .then((_) => {
          // handle success //
          const { loadedStores } = state.storeState;
          const storeDropdownData: StoreDropdownData[] = loadedStores.map((store) => {
            return {
              storeId: store._id,
              storeName: store.title
            }
          });
          setStoreDropDownData(storeDropdownData);
          checkSetValues(currentStoreItemData) ? setNewForm(false) : setNewForm(true);
        })
        .catch((_) => {
          // handle error //
        });
      }
    return () => { mounted = false };
  }, []);
  useEffect(() => {
    if (!storeItemFormOpen) {
      window.scrollTo({ top: 0, behavior: "smooth" });
    }
  }, [ storeItemFormOpen ]);
  
  useEffect(() => {
    checkSetValues(currentStoreItemData) ? setNewForm(false) : setNewForm(true);
  }, [ currentStoreItemData]);
  // component return //
  return (
    <div id="storeItemFormHolder">
      { loading ? <LoadingBar /> : null }
      <FormErrorComponent error={error as AxiosError} />
      {
        !newForm ?
          <React.Fragment>
            <Grid.Row>
              <Grid.Column mobile={16} tablet={14} computer={14}>
                <h1>Details</h1>
                <div className="storeItemFormHolderDetails">
                  <h3>Store Item name:</h3>
                  <p>{name}</p>
                </div>
                <div className="storeItemFormHolderPrice">
                  <h3>Store Item price:</h3>
                  <p>{price}</p>
                </div>
                <div className="storeItemFormHolderDetails">
                  <h3>Store Item description:</h3>
                  <p>{description}</p>
                </div>
                <div className="storeItemFormHolderDetails">
                  <h3>Store Item details:</h3>
                  <p>{details}</p>
                </div>
                <div className="storeItemFormHolderCategories">
                  <h3>Listed Store Item categories: </h3>
                  { categories.length === 0 ? "no categories listed" : null}
                  <List horizontal>
                    {
                      categories.map((category) => {
                        return (
                          <List.Item key={category}>
                            <List.Content>
                              <List.Header>{category}</List.Header>
                            </List.Content>
                          </List.Item>
                        )
                      })
                    }
                  </List>
                </div>
                <div className="storeItemFormHolderTimestamps">
                  <span>Created At: <strong>{ConvertDate.international(createdAt)}</strong></span>
                  <span>Edited At: <strong>{ConvertDate.international(editedAt)}</strong></span>
                </div>
              </Grid.Column>
            </Grid.Row>
            <Grid.Row>
              <Grid.Column mobile={16} tablet={14} computer={14}>
                <StoreItemImgPreviewHolder state={state} dispatch={dispatch} />
              </Grid.Column>
            </Grid.Row>
            <StoreItemImgUplForm />
          </React.Fragment>
          : null
      }
      <Grid.Row>
        <Grid.Column mobile={16} tablet={15} computer={14}>
          <Button  id="storeItemFormToggleBtn" 
            onClick={toggleForm} 
            content={ !storeItemFormOpen ? "Open Form" : "Close Form"}
          />
          {
            storeItemFormOpen ? 
              <StoreItemForm 
                storeId={storeId}
                storeName={storeName}
                name={name}
                price={price}
                details={details}
                description={description}
                categories={categories}
                newForm={newForm}
                activeStores={storeDropDownData ? storeDropDownData : []}                
                handleCreateStoreItem={handleCreateStoreItem}
                handleUpdateStoreItem={handleUpdateStoreItem}
              /> 
            : null
          }
        </Grid.Column>
      </Grid.Row>
    </div>
  );
};

export default withRouter(StoreItemFormHolder);

