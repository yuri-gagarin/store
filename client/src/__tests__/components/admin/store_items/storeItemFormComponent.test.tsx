import React from "react"
import sinon from "sinon";
import { createStoreItem } from "../../../../components/admin_components/store_items/actions/APIStoreItemActions"
import { mount, shallow, ReactWrapper } from "enzyme";
import { Button } from "semantic-ui-react";
// component imports //
import StoreItemFormHolder from "../../../../components/admin_components/store_items/forms/StoreItemFormHolder";
import StoreItemForm from "../../../../components/admin_components/store_items/forms/StoreItemForm";
import StoreItemImageUplForm from "../../../../components/admin_components/store_items/forms/StoreItemImgUplForm";
import StoreItemImgPreviewHolder from "../../../../components/admin_components/store_items/image_preview/StoreItemImgPreviewHolder";
import StoreItemImgPreviewThumb from "../../../../components/admin_components/store_items/image_preview/StoreItemImgThumb";
import LoadingBar from "../../../../components/admin_components/miscelaneous/LoadingBar";
// helpers //
import { generateCleanState } from "../../../../test_helpers/miscHelpers";
import { createMockStoreItems, setMockStoreItemState } from "../../../../test_helpers/storeItemHelpers";
import { IGlobalAppContext, IGlobalAppState, StateProvider } from "../../../../state/Store";
import { act } from "react-dom/test-utils";
import moxios from "moxios";
import { createMockStores } from "../../../../test_helpers/storeHelpers";

const extractContext = (): IGlobalAppContext => {
  type WrapperProps = {
    value: IGlobalAppContext;
  }
  let wrapper = shallow<WrapperProps>(<StateProvider />)
  const { state, dispatch } = wrapper.props().value;
  return {
    state,
    dispatch
  };
}
const ConnectToContext = (component: React.Component): IGlobalAppContext => {
  type WrapperProps = {
    value: IGlobalAppContext;
  }
  let wrapper = shallow<WrapperProps>(<StateProvider />)
  const { state, dispatch } = wrapper.props().value;
  return {
    state,
    dispatch
  };
}
describe("StoreItemFormHolder Component tests", () => {
  let wrapper: ReactWrapper; 
  
  describe("Default Form Holder state",  () => {

    beforeAll(() => {
      window.scrollTo = jest.fn();
      wrapper = mount(<StoreItemFormHolder />);
    });

    it("Should Properly Mount Form Holder", () => {
      expect(wrapper.html()).toBeDefined();
    });
    it("Form Should be closed by default", () => {
      const form = wrapper.find(StoreItemForm);
      expect(form.length).toEqual(0);
    });
    it("Should have a Form toggle Button", () => {
      const toggleButton = wrapper.find(Button);
      expect(toggleButton.length).toEqual(1);
    })
  });

  describe("Form Holder state OPEN - NO Current StoreItem Data",  () => {

    beforeAll(() => {
      window.scrollTo = jest.fn();
      wrapper = mount(<StoreItemFormHolder />);
    });

    it("Should Properly Mount Form Holder", () => {
      const toggleButton = wrapper.find("#storeItemFormToggleBtn");
      toggleButton.at(0).simulate("click")
      // open button clicked //
      expect(wrapper.html()).toBeDefined();

    });
    it("Should have a Form toggle Button", () => {
      const toggleButton = wrapper.render().find('#storeItemFormToggleBtn');
      expect(toggleButton.length).toEqual(1);
    });
    it("Should have a Form Create Button", () => {
      const toggleButton = wrapper.render().find('#adminStoreItemFormCreate');
      expect(toggleButton.length).toEqual(1);
    });
    it("Should have the Form rendered after toggle button", () => {
      const form = wrapper.find(StoreItemForm);
      expect(form.length).toEqual(1);
    });
    it("Should NOT have the Image Preview rendered", () => {
      const imgPreviewHolder = wrapper.find(StoreItemImgPreviewHolder);
      expect(imgPreviewHolder.length).toEqual(0);
    });
    it("Should NOT have the Image Upload Form rendered", () => {
      const imgUploadForm = wrapper.find(StoreItemImageUplForm);
      expect(imgUploadForm.length).toEqual(0);
    });

  });
  /*
  describe("Form Holder state OPEN - WITH Current StoreItem Data - NO IMAGES",  () => {
    let state: IGlobalAppState;
    beforeAll(() => {
      window.scrollTo = jest.fn();
      state = setMockStoreItemState({ currentStoreItem: true });
      wrapper = mount(<StoreItemFormHolder />);
    });

    it("Should Properly Mount Form Holder", () => {
      expect(wrapper).toMatchSnapshot();
    });
    it("Should have a Form toggle Button", () => {
      const toggleButton = wrapper.render().find('#storeItemFormToggleBtn');
      expect(toggleButton.length).toEqual(1);
    });
    it("Should have the Form rendered", () => {
      const form = wrapper.find(StoreItemForm);
      expect(form.length).toEqual(1);
    });
    it("Should have a Form Update Button", () => {
      const toggleButton = wrapper.render().find('#adminStoreItemFormUpdate');
      expect(toggleButton.length).toEqual(1);
    });
    it("Should have the Image Preview rendered", () => {
      const imgPreviewHolder = wrapper.find(StoreItemImgPreviewHolder);
      expect(imgPreviewHolder.length).toEqual(1);
    });
    it("Should NOT render any preview images", () => {
      const previewThumb = wrapper.find(StoreItemImgPreviewThumb);
      expect(previewThumb.length).toEqual(0);
    })
    it("Should have the Image Upload Form rendered", () => {
      const imgUploadForm = wrapper.find(StoreItemImageUplForm);
      expect(imgUploadForm.length).toEqual(1);
    });

  });
  
  describe("Form Holder state OPEN - WITH Current StoreItem Data - WITH IMAGES",  () => {
    let state: IGlobalAppState;
    beforeAll(() => {
      window.scrollTo = jest.fn();
      state = setMockStoreItemState({ currentStoreItem: true, storeItemImages: 3 });
      wrapper = mount(<StoreItemFormHolder />);
    });

    it("Should Properly Mount Form Holder", () => {
      expect(wrapper.html()).toBeDefined();
    });
    it("Should have a Form toggle Button", () => {
      const toggleButton = wrapper.render().find('#storeItemFormToggleBtn');
      expect(toggleButton.length).toEqual(1);
    });
    it("Should have the Form rendered", () => {
      const form = wrapper.find(StoreItemForm);
      expect(form.length).toEqual(1);
    });
    it("Should have a Form Update Button", () => {
      const toggleButton = wrapper.render().find('#adminStoreItemFormUpdate');
      expect(toggleButton.length).toEqual(1);
    });
    it("Should have the Image Preview rendered", () => {
      const imgPreviewHolder = wrapper.find(StoreItemImgPreviewHolder);
      expect(imgPreviewHolder.length).toEqual(1);
    });
    it("Should render a correct number of preview images", () => {
      const previewThumb = wrapper.find(StoreItemImgPreviewThumb);
      const numberOfImages = state.storeItemState.currentStoreItemData.images.length;
      expect(previewThumb.length).toEqual(numberOfImages);
    })
    it("Should have the Image Upload Form rendered", () => {
      const imgUploadForm = wrapper.find(StoreItemImageUplForm);
      expect(imgUploadForm.length).toEqual(1);
    });
  });
  */
  describe("Form Holder state OPEN - MOCK Submit action",  () => {
    let state: IGlobalAppState; let wrapper: ReactWrapper;

    beforeAll( async () => {
      window.scrollTo = jest.fn();
      moxios.install();

      wrapper = mount(
        <StateProvider>
          <StoreItemFormHolder />
        </StateProvider>
      );
    
      await act( async () => {
        moxios.stubRequest("/api/stores", {
          status: 200,
          response: {
            responseMsg: "OK",
            stores: createMockStores(5)
          }
        });
      });
      moxios.uninstall()
    });
    it("Should have a submit button", () => {
      wrapper.update();
      wrapper.find("#storeItemFormToggleBtn").at(0).simulate("click").update();
      const adminStoreItemFormCreate = wrapper.find("#adminStoreItemFormCreate").at(0);
      expect(adminStoreItemFormCreate.length).toEqual(1)
    })
    
    it("Should handle the 'handleCreateStoreItemAction, show 'LoadingBar' Component", async () => {
      moxios.install();
      await act( async () => {
        moxios.stubRequest("/api/store_items/create", {
          status: 200,
          response: {
            responseMsg: "All Good",
            newStoreItem: createMockStoreItems(1)[0]
          }
        });
        const adminStoreItemFormCreate = wrapper.find("#adminStoreItemFormCreate").at(0);
        adminStoreItemFormCreate.simulate("click");
        //expect(wrapper.find(LoadingBar).length).toEqual(1);
      });
      // expect(sinon.spy(createStoreItem)).toHaveBeenCalled()

    });
    it("Should NOT show the 'LoadingBar' Component after successful API call", () => {
      expect(wrapper.find(LoadingBar).length).toEqual(0);
    });
    
    it("Should NOT show the 'StoreItemForm' Component after successful API call", () => {
      act(() => {
        wrapper.update();
      })
      expect(wrapper.find(StoreItemForm).length).toEqual(0);
    });
    
  });

});