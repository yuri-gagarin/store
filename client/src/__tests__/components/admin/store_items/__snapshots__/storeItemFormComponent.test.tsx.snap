// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StoreItemFormHolder Component tests Form Holder state OPEN - WITH Current StoreItem Data - NO IMAGES Should Properly Mount Form Holder 1`] = `
<StoreItemFormHolder>
  <div
    id="storeItemFormHolder"
  >
    <FormErrorComponent
      error={null}
    />
    <GridRow>
      <div
        className="row"
      >
        <GridColumn
          computer={14}
          mobile={16}
          tablet={15}
        >
          <div
            className="fourteen wide computer sixteen wide mobile fifteen wide tablet column"
          >
            <Button
              as="button"
              content="Open Form"
              id="storeItemFormToggleBtn"
              onClick={[Function]}
            >
              <Ref
                innerRef={
                  Object {
                    "current": <button
                      class="ui button"
                      id="storeItemFormToggleBtn"
                    >
                      Open Form
                    </button>,
                  }
                }
              >
                <RefFindNode
                  innerRef={
                    Object {
                      "current": <button
                        class="ui button"
                        id="storeItemFormToggleBtn"
                      >
                        Open Form
                      </button>,
                    }
                  }
                >
                  <button
                    className="ui button"
                    id="storeItemFormToggleBtn"
                    onClick={[Function]}
                  >
                    Open Form
                  </button>
                </RefFindNode>
              </Ref>
            </Button>
          </div>
        </GridColumn>
      </div>
    </GridRow>
  </div>
</StoreItemFormHolder>
`;

exports[`StoreItemFormHolder Component tests Form Holder state OPEN - WITH Current StoreItem Data - WITH IMAGES - Submit action Should have a submit button 1`] = `
<Button
  as="button"
  content="Create  New Store Item"
  id="adminStoreItemFormCreate"
  onClick={[Function]}
  type="submit"
>
  <Ref
    innerRef={
      Object {
        "current": <button
          class="ui button"
          id="adminStoreItemFormCreate"
          type="submit"
        >
          Create  New Store Item
        </button>,
      }
    }
  >
    <RefFindNode
      innerRef={
        Object {
          "current": <button
            class="ui button"
            id="adminStoreItemFormCreate"
            type="submit"
          >
            Create  New Store Item
          </button>,
        }
      }
    >
      <button
        className="ui button"
        id="adminStoreItemFormCreate"
        onClick={[Function]}
        type="submit"
      >
        Create  New Store Item
      </button>
    </RefFindNode>
  </Ref>
</Button>
`;
